### ISSUE TITLE
New syntax for state and params

### ISSUE THREAD

**Issue by catmando** [hyperstack-org/hyperstack#39]
Take this example from the new website:

```ruby
class UsingState < Hyperloop::Component

  state :show  # true means show the input field, false means hide it
  state :value # current value of input field as user types

  render(DIV) do
    button.on(:click) { mutate.show !state.show }
    DIV { example } if state.show
  end

  def button
    BUTTON(class: 'ui primary button') do
      state.show ? "Hide" : "Show"
    end
  end

  def example
    BR()
    DIV(class: 'ui input') do
      INPUT(type: :text).on(:change) do |e|
        mutate.value e.target.value
      end
    end

    H3 { "#{state.value}" }
  end
end
```

Right now we consider each store and component to have a "state" object. 
You access individual states in that object using the `state` and `mutate` methods.

This follows react closely, but from a ruby/object oriented perspective its backwards.

Rather we should consider each "state" to **be** an object, which has reader, writer and mutator (and possibly other) methods associated.

Here is the same code written that way:

```ruby
class UsingState < Hyperloop::Component

  state :show  # true means show the input field, false means hide it
  state :value # current value of input field as user types

  render(DIV) do
    button.on(:click) { show.state = !show.state }
    DIV { example } if show.state
  end

  def button
    BUTTON(class: 'ui primary button') do
      show.state ? "Hide" : "Show"
    end
  end

  def example
    BR()
    DIV(class: 'ui input') do
      INPUT(type: :text).on(:change) do |e|
        value.state = e.target.value
      end
    end

    H3 { "#{value.state}" }
  end
end
```

In detail:

currently there are basically three methods associated with each state:

a reader:  `state.show` for example
a writer: `mutate.show(false)`
a mutator: `mutate.show` <- no parameter switches behavior under the hood.

New approach:

the reader: `show.state`
the writer:   `show.state=`
the mutator: `show.mutate`

Other advantages:

now that the state is the object, it can handy methods associated cleanly.  For example we can add `set?` and `toggle!` methods to nicely deal with boolean states:

```ruby
class UsingState < Hyperloop::Component

  state :show  # true means show the input field, false means hide it
  state :value  # current value of input field as user types

  render(DIV) do
    button.on(:click) { show.toggle! }
    DIV { example } if show.set?
  end

  def button
    BUTTON(class: 'ui primary button') do
      show.set? ? "Hide" : "Show"
    end
  end

  def example
    BR()
    DIV(class: 'ui input') do
      INPUT(type: :text).on(:change) do |e|
        value.state = e.target.value
      end
    end

    H3 { "#{value.state}" }
  end
end
```

What about the mutator???

When dealing with scalar values like the above example you don't "mutate" the state, but rather just assign a whole new value to it.  In cases where the state is not a scalar (like an array, hash or other object) and you want to mutate it, then indeed you would say `list_of_items.mutate << new_item`. 

Overall its much clearer, and under the hood the implementation will be much cleaner and more effecient.

Downside?

**You have to declare the states.**   Currently you can just use a state on the fly and it gets created inside the state object.  In the above approach the system has to know before hand that the state exists.   The good news is this means if you mistype the name of a state then you would get the normal method missing error.

**Not symetrical with params.**  I would at least like to try also getting rid of the params object and just declaring methods for each param.  But its a separate issue.

Next Steps?

Both systems can coexist, so its a matter of providing a mixin.  The question is how to try this out:

a.  Provide a mixin that adds the new behavior
b.  Build the new behavior in, and provide a mixin for the old behavior.

I think the choice will depend on how people vote for the above.  If most people like it, we go with (a), if people are concerned we go with (b).

So please THUMBS UP/ THUMBS DOWN (if you have to :-) ) and comment...

**Comment by barriehadfield** [comment]
@catmando I think its an improvement overall so I vote for B as we will only document one way of working. It is odd though, State becomes very close to a Store.

**Comment by djshitcoin** [comment]
What is the difference between the writer and mutator, don't they do the same thing?

how about just `state.show` for the reader (as before) and `state.show=` for the writer, and that's it.

**Comment by catmando** [comment]
@djshitcoin its a subtle difference.   Consider this:

```ruby
  list_of_users.state = []  
  # users is now an empty array, and because state= is a separate operator
  # the system can be notified that the state of list_of_users has changed.
  
  list_of_users.state  # returns that array

  list_of_users.state << "@djshitcoin"
  # this READS the state of list_of_users, which is still the same empty array
  # then it shoves a new value on the end of the array
  # but there is no way that the system can now that the state has actually
  # changed because we have "mutated" the structure of the array, but we did
  # not change the value of the state (its still the same array)

  # instead we need the mutate method:
  list_of_users.mutate << "@djshitcoin"

  # the value returned by mutate is not actually the array but a wrapper that acts just like the array
  # but when ever changes are made the system is notified that the contents has changed.

**Comment by catmando** [comment]
FYI the other big picture solution to this is to use immutable data.   That means you never actually mutate the data, but instead you make a copy, change the copy, and assign the new copy to the state.

This is not as odd as it seems as we naturally do this with strings all the time:

```ruby
list_of_users.state = ""
list_of_users.state += " @djshitcoin" # += just uses = and + separately under the hood

# like wise if you wanted to, you could do the same with arrays:

list_of_users.state = []
list_of_users.state += ['@djshitcoin']
```

**Comment by sfcgeorge** [comment]
* I always declare my states anyway, I think the method missing errors will be helpful.
* The old syntax did feel a bit reversed, I like this better.
* I think I prefer the `+=` way and would be in favour of getting rid of mutate.

```ruby
mutate.counter state.counter + 1
counter.state = counter.state + 1
counter.state += 1 # will work anyway I think, just syntax
```

But if we're defining methods why not just define `xyz` and `xyz=` directly? No need for repetetive `xyz.state`.

```ruby
counter # reader
counter = 1 # writer
counter += 1 # syntax
counter = !counter # toggle
counter? # boolean
```

But what about params? Well, why don't they just get the reader but no writer. How can you tell what's a param and what's state? 1) Who cares, you can read from both just the same. 2) If you try writing to a param you'll get method missing and figure it out, OR we could define the writer to raise a helpful message. 3) You declare them all at the top, easy to see which is which.

```ruby
class DescriptionToggle < Hyperloop::Component
  state :show  # true means show description, false means hide it
  param :value # value of expandable description paragraph

  render(DIV) do
    button.on(:click) { show = !show }
    P { value } if show?
  end

  def button
    BUTTON(class: 'ui primary button') do
      show ? "Hide" : "Show"
    end
  end
end
```

**Comment by sfcgeorge** [comment]
Damn you're right, I can't believe I made that mistake. Updated my example in case anyone else sees it, obviously less appealing now.

I do like that idea from the perspective of parity with `attr_*` methods, but agree it will/did trip people up (even those who know you have to use self 😑).

Then yes, I'm in favour of:

```ruby
param.foo
state.bar
state.bar = 42
```

There's still some nicely improved parity there with params and state.

**Comment by catmando** [comment]
ahhh that is almost the way it works today... proposal is basically to flip around things
```
  params :foo
  state :bar
...
  foo # param reader 
  bar.state # read the state of bar
  bar.state = [] # rewrite the state of bar
  bar.mutate << 42  # mutate bars's current state
```

**Comment by catmando** [comment]
so final status (implemented in 1.0.alpha1) is:

States and params are both presented as instance variables within a component.  The same goes for states inside of a store.

For params the instance variable name is the param name `CamelCased`.  This allows params to clearly stand out when reading code.

To indicate that a state has been mutated you use the `mutate` method:

`mutate @my_state = !@my_state`

mutate takes any number of args or a block so can be used in a number of forms all which will have the same result:

```ruby
mutate @state1 = 12, @state2 = 13
# or
mutate do
  @state1 = 12
  @state2 = 13
end
# or
@state1 = 12
@state2 = 13
mutate
```

Any class can include `Hyperstack::State::Observable` which adds the observe and mutate methods, plus other helpers.  The methods are added to both the instances and the class (i.e. the eigenclass).

The basic methods are `observe` and `mutate`.  The `observe` method indicates that this object's internal state is being *observed* and so the current rendering component will be added to a list of observers of this object.

The `mutate` method indicates the object's internal state is changing, and so any components who observed the current state need to be rerendered.

Additional helper methods are built on these methods:

`mutator` works like `define_method` but wraps the resulting method in a `mutate` call.
`observer` likewise wraps the resulting method in an `observer` call.

`state_reader, state_writer` and `state_accessor` work like their attribute counter parts but wrap observe/mutate the instance variable as appropriate.

Components also inherit `Hyperstack::State::Observable`.  Components always "observe" their own state automatically, but you still need to call `mutate` when you change state.
